packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts

```diff
@@ -0,0 +1,444 @@
+import { DebugLogger } from '@affine/debug';
+import type { AffineTextAttributes } from '@blocksuite/blocks';
+import type { DeltaInsert } from '@blocksuite/inline';
+import type { Job, WorkspaceService } from '@toeverything/infra';
+import {
+  Document,
+  Entity,
+  IndexedDBIndexStorage,
+  IndexedDBJobQueue,
+  JobRunner,
+} from '@toeverything/infra';
+import { difference } from 'lodash-es';
+import { Array as YArray, Map as YMap, type Text as YText } from 'yjs';
+import { applyUpdate, Doc as YDoc } from 'yjs';
+
+import {
+  type BlockIndexSchema,
+  blockIndexSchema,
+  docIndexSchema,
+} from '../schema';
+
+const logger = new DebugLogger('crawler');
+
+interface IndexerJobPayload {
+  docId: string;
+}
+
+export class DocsIndexer extends Entity {
+  private readonly jobQueue = new IndexedDBJobQueue<IndexerJobPayload>(
+    'jq:' + this.workspaceService.workspace.id
+  );
+
+  private readonly runner = new JobRunner(
+    this.jobQueue,
+    (jobs, signal) => this.execJob(jobs, signal),
+    () => new Promise<void>(resolve => requestIdleCallback(() => resolve()))
+  );
+
+  private readonly indexStorage = new IndexedDBIndexStorage(
+    'idx:' + this.workspaceService.workspace.id
+  );
+
+  readonly docIndex = this.indexStorage.getIndex('doc', docIndexSchema);
+
+  readonly blockIndex = this.indexStorage.getIndex('block', blockIndexSchema);
+
+  private readonly workspaceEngine = this.workspaceService.workspace.engine;
+
+  private readonly workspaceId = this.workspaceService.workspace.id;
+
+  constructor(private readonly workspaceService: WorkspaceService) {
+    super();
+  }
+
+  setupListener() {
+    this.workspaceEngine.doc.storage.eventBus.on(event => {
+      if (event.clientId === this.workspaceEngine.doc.clientId) {
```

Comment:
how about remote updates?

Reply:
This also includes the remote remote, where the clientid represents the client that receives the remote update. When there are multiple tabs, the tab that receives the update is responsible for pushing it into the job queue.